### PR TITLE
Added conversion for cpu.weight v2

### LINKED ISSUE #2166: cgroup2: cpu.shares is broken
[`cpu_v2.go` supports libcontainer `CpuWeight`](https://github.com/opencontainers/runc/blob/46def4cc4cb7bae86d8c80cedd43e96708218f0a/libcontainer/cgroups/fs/cpu_v2.go#L45), but there is no code that translates OCI `cpu.shares` to libcontainer `CpuWeight`. So OCI `cpu.shares` is currently just ignored.

The translation rule should follow crun convention: https://github.com/containers/crun/blob/master/crun.1.md#cpu-controller

### PR DESCRIPTION
**Pull request by Zyqsempai**
Fix #2166 

Added conversion for cpu.weight v2 from cpu.shares

Signed-off-by: Boris Popovschi <zyqsempai@mail.ru>

### CODE REVIEW

**Comment by Zyqsempai** [comment]
@crosbymichael @cyphar @AkihiroSuda PTAL

**Comment by AkihiroSuda** [comment]
@mrunalp @hqhq @dqminh PTAL (and https://github.com/opencontainers/runc/pull/2212 https://github.com/opencontainers/runc/pull/2192 as well as this one)

**Comment by mrunalp** [comment]
LGTM

File: libcontainer/specconv/spec_linux.go [hunk 1]
```diff
@@ -405,6 +405,9 @@ func CreateCgroupConfig(opts *CreateOpts) (*configs.Cgroup, error) {
 		if r.CPU != nil {
 			if r.CPU.Shares != nil {
 				c.Resources.CpuShares = *r.CPU.Shares
+
+				//CpuWeight is used for cgroupv2 and should be converted
+				c.Resources.CpuWeight = (1 + ((*r.CPU.Shares-2)*9999)/262142)
```
Comment (hqhq) [inline libcontainer/specconv/spec_linux.go]: We should do this with a function so people know it's converting runtime-spec items to cgroupv2 value.

And comments like "convert from [2-262144] to [1-10000]" would be helpful, more explaination such as value 262144 comes from Linux kernel definition "#define MAX_SHARES               (1UL << 18)" will also help.

Comment (h-vetinari) [reply libcontainer/specconv/spec_linux.go]: These have been collected by @giuseppe for crun [here](https://github.com/giuseppe/crun/blob/master/crun.1.md#cgroup-v2) already. I'm thinking about how to go about adding these to the runtime-spec as per the discussion in opencontainers/runtime-spec#1002.

Should the functions you mention be part of the spec already?

**Comment by Zyqsempai** [comment]
@hqhq Done, PTAL

File: update.go [hunk 2]
```diff
@@ -302,3 +304,12 @@ other options are ignored.
...
 }
+
+// Since the OCI spec is designed for cgroup v1, in some cases
+// there is need to convert from the cgroup v1 configuration to cgroup v2
+// the formula for cpuShares is y = (1 + ((x - 2) * 9999) / 262142)
+// convert from [2-262144] to [1-10000]
+// 262144 comes from Linux kernel definition "#define MAX_SHARES (1UL << 18)"
+func convertCPUSharesToCgroupV2Value(cpuShares uint64) uint64 {
+	return (1 + ((cpuShares-2)*9999)/262142)
+}
```
Comment (hqhq) [inline update.go]: All these convertion functions should be put in a utils file so it'll be easier to remove them when we have these items in runtime-spec someday.

I've reclaimed my LGTM in the other PR.

Comment (Zyqsempai) [reply update.go]: Make sense!

**Comment by Zyqsempai** [comment]
@hqhq PTAL on more time here as well.

File: libcontainer/utils/utils.go [hunk 3]
```diff
@@ -110,3 +110,12 @@ func Annotations(labels []string) (bundle string, userAnnotations map[string]str
...
 }
+
+// Since the OCI spec is designed for cgroup v1, in some cases
+// there is need to convert from the cgroup v1 configuration to cgroup v2
+// the formula for cpuShares is y = (1 + ((x - 2) * 9999) / 262142)
+// convert from [2-262144] to [1-10000]
+// 262144 comes from Linux kernel definition "#define MAX_SHARES (1UL << 18)"
+func ConvertCPUSharesToCgroupV2Value(cpuShares uint64) uint64 {
+	return (1 + ((cpuShares-2)*9999)/262142)
+}
```
Comment (hqhq) [inline libcontainer/utils/utils.go]: Sorry I should be more specific, but maybe libcontainer/cgroups/utils.go is a better place?

**Comment by hqhq** [comment]
LGTM

**Comment by AkihiroSuda** [comment]
merge?

**Comment by hqhq** [comment]
Needs re-approve. @mrunalp

**Comment by mrunalp** [comment]
LGTM

**Pull request merged**